### PR TITLE
Closes #1822 implement derive_param_extreme_record

### LINKED ISSUE #1822: Feature Request: Implement `derive_param_extreme_record()`
### Feature Idea

Implement a function selecting the first or last record from multiple sources and adding the selected record as a new records to the input dataset.

### Relevant Input

_No response_

### Relevant Output

_No response_

### Reproducible Example/Pseudo Code

```
derive_param_extreme_record(
  adevent,
  sources = list(
    records(
      dataset_name = "cm",
      filter = CMDECOD == "ANTI-CANCER THERAPY",
      new_vars = exprs(
        ADT = convert_dtc_to_dt(CMSTDTC),
        AVALC = CMDECOD),
    ),
    records(
      dataset_name = "pr",
      filter = PRDECOD == "ANTI-CANCER SURGERY",
      new_vars = exprs(
        ADT = convert_dtc_to_dt(PRSTDTC),
        AVALC = PRDECOD),
    ),
  source_datasets = list(cm = cm, pr = pr),
  order = exprs(ADT),
  mode = "first",
  set_values_to = exprs(
    PARAMCD = "FIRSTACT",
    PARAM = "First anti-cancer therapy"
  )
)
```

### PR DESCRIPTION
**Pull request by zdz2101**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiraldev/devel/articles/development_process.html) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `devel` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/devel/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/devel/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/devel/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/cran-release/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately 
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

File: R/derive_param_extreme_record.R [hunk 1]
```diff
@@ -0,0 +1,233 @@
+#' Adds a Parameter based on first or last record from multiple sources
+#'
+#' Adds a Parameter based on first or last record from multiple sources
```
Comment (bms63) [inline R/derive_param_extreme_record.R:3]: Perhaps the second second isn't need, if the function is self-explanatory??

File: R/derive_param_extreme_record.R [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
```
Comment (bms63) [inline R/derive_param_extreme_record.R:1]:
```suggestion
#' Adds a Parameter Based on First or Last Record from Multiple Sources
```
We've been consistently capitalizing most words that aren't articles in our function titiles.

File: R/derive_param_extreme_record.R [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+      order = order,
+      mode = mode
+    ) %>%
+    mutate(!!!set_values_to)
+
+  # Bind the parameter rows back to original dataset
+  data <- bind_rows(dataset, param_data)
+  return(data)
+}
+#' Create a `records_source` Object
```
Comment (bms63) [inline R/derive_param_extreme_record.R:186]: Can we insert a blank line here

File: tests/testthat/test-derive_param_extreme_record.R [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+      PARAMCD = "FIRSTACT",
+      PARAM = "First Anti-Cancer Therapy"
+    )
+  )
+
+  expect_dfs_equal(expected_output, actual_output, keys = c("USUBJID", "PARAMCD", "PARAM", "ADT", "AVALC")) # nolint
+})
+
+## Test 2: Error order variable not inside source datasets ----
+test_that("derive_param_extreme_record Test 2: Error order variable not inside source datasets", {
```
Comment (bms63) [inline tests/testthat/test-derive_param_extreme_record.R:65]:
```suggestion
test_that("derive_param_extreme_record Test 2: Error given when order variable is not inside source datasets", {
```

**Review by bms63 — COMMENTED** [review]
Do the tests cover the function 100%?

Looking really good - I will take look again tomorrow, but just wanted to give some preliminary feedback.

File: tests/testthat/test-derive_param_extreme_record.R [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    "1001",        "3", "2023-01-01",    "TST", "TEST"
+  )
+
+  pr <- tibble::tribble(
+    ~STUDYID, ~USUBJID, ~PRDECOD,     ~PRSTDTC,
+    "1001",        "1",    "ACS", "2020-12-27",
+    "1001",        "2",    "ACS", "2021-12-25",
+    "1001",        "3",    "ACS", "2022-12-25",
+  )
+  expect_error(
```
Comment (bundfussr) [inline tests/testthat/test-derive_param_extreme_record.R:132]: Please add `regexp` or `class` argument to verify the the expected error is issued.

File: R/derive_param_extreme_record.R [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+#'   source_datasets = list(cm = cm, pr = pr),
+#'   by_vars = exprs(USUBJID),
+#'   order = exprs(ADT),
+#'   mode = "first",
+#'   set_values_to = exprs(
+#'     PARAMCD = "FIRSTACT",
+#'     PARAM = "First Anti-Cancer Therapy"
+#'   )
+#' )
+derive_param_extreme_record <- function(dataset,
```
Comment (bundfussr) [inline R/derive_param_extreme_record.R:139]: `dataset` argument should be defaulted to `NULL`.

File: R/derive_param_extreme_record.R (hunk 1 above)
Comment (zdz2101) [reply R/derive_param_extreme_record.R:3]: image
feels like our convention always seems to have both available, second can be a little more "descriptive" which I certainly think we can make slightly more clear

File: R/derive_param_extreme_record.R [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+      by_vars = by_vars,
+      order = order,
+      mode = mode
+    ) %>%
+    mutate(!!!set_values_to)
+
+  # Bind the parameter rows back to original dataset
+  data <- bind_rows(dataset, param_data)
+  return(data)
+}
```
[the remark this answers is not in the document]
Comment (zdz2101) [reply R/derive_param_extreme_record.R:185]: I'll play around with it

File: R/derive_param_extreme_record.R (hunk 1 above)
Comment (bms63) [reply R/derive_param_extreme_record.R:3]: Oh yeah - I think we should not just repeat the title in that section.  Perhaps another clean up task to give more context, assumption, etc in that section

**Comment by kaz462** [comment]
There are four functions for deriving extreme event/record as a new parameter, the function names are a bit confusing to me when I try to compare them: `_record(s)`/`_event`, with/without `_param`


  | Data source:      one or multiple | Extreme is ordered by - | filter/dataset source argument name
-- | -- | -- | -- 
1.[`derive_extreme_records`](https://pharmaverse.github.io/admiral/cran-release/reference/derive_extreme_records.html)| one | `order` argument | `filter`/ |
2.[`derive_extreme_event`](https://pharmaverse.github.io/admiral/cran-release/reference/derive_extreme_event.html)  | one | the order of `events` entered and the `order` argument |
3.[`derive_param_extreme_event`](https://pharmaverse.github.io/admiral/cran-release/reference/derive_param_extreme_event.html) | one  |`order` argument| `filter_source`/`dataset_source`
4.`derive_param_extreme_record` | multiple   | the order of `sources` entered and the `order` argument   | /`source_datasets` |



* What do you think of combining some functions? e.g., 

     * function 3 is a special case of function 2 

     * function 4 is an extension of function 2 by allowing multiple data sources  

* Should we make some argument name consisnent across those functions?

**Comment by kaz462** [comment]
> What do you think of combining some functions? e.g.,

Sorry, I just realized function 1 and 3 has been combined through https://github.com/pharmaverse/admiral/pull/1882
In this case, should we combine function 2 and 4?

**Comment by bundfussr** [comment]
@kaz462 , good point!
We had a first clean-up round at the beginning of this year. Once we have implemented everything from the first round, we can have a second one. I have created an issue to collect suggestions for clean up (https://github.com/pharmaverse/admiral/issues/1904) and added your comment there.

For `derive_extreme_event()` I would expect updates required by admiralonco.

**Comment by bms63** [comment]
@kaz462 it would be great to get you involved in the function clean-up/re-alignment.  We were meeting every two weeks to discuss a list that @bundfussr made to help us review.  It seems like you might have your own list too! 

I think we were going to restart our clean-up/re-alignment after the release? @bundfussr shall we get these back on the menu?

**Comment by kaz462** [comment]
@bundfussr Thanks for creating a new issue for clean up! @bms63 I've only got this list so far, but happy to be involved and discuss further!

I just edited the summary by adding the last point about #1725, also updated in https://github.com/pharmaverse/admiral/issues/1904
> * function 3 is a special case of function 2
> * function 4 is an extension of function 2 by allowing multiple data sources
> * function 1 and 3 has been combined through [Closes #1725 Combine derive_extreme_records() and derive_param_extreme_event() #1882](https://github.com/pharmaverse/admiral/pull/1882)

**Comment by zdz2101** [comment]
@bms63 @bundfussr I think all comments/feedback have been addressed

File: R/derive_param_extreme_record.R [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+#'
+#' @param by_vars By variables
+#'
+#'   If the parameter is specified, for each by group the observations are
+#'   selected separately.
+#'
+#' @param order Sort order
+#'
+#'   If the argument is set to a non-null value, for each by group the first or
+#'   last observation from the additional dataset is selected with respect to
```
[the remark this answers is not in the document]
Comment (bms63) [reply R/derive_param_extreme_record.R:27]: @zdz2101 is this one that you need to update or not going to??

File: R/derive_param_extreme_record.R [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+#' @param set_values_to Variables to be set
+#'
+#'   The specified variables are set to the specified values for the new
+#'   observations.
+#'
+#'   A list of variable name-value pairs is expected.
+#'   + LHS refers to a variable.
+#'   + RHS refers to the values to set to the variable. This can be a string, a
+#'   symbol, a numeric value or `NA`, e.g., `exprs(PARAMCD = "TDOSE", PARCAT1 =
+#'   "OVERALL")`. More general expression are not allowed.
```
[the remark this answers is not in the document]
Comment (bms63) [reply R/derive_param_extreme_record.R:56]: Another one to look

File: R/derive_param_extreme_record.R [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+#' @keywords source_specifications
+#' @family source_specifications
+#'
+#' @seealso [derive_param_extreme_record()]
+#'
+#' @return  An object of class `records_source`
+#' @export
+records_source <- function(dataset_name,
+                           filter = NULL,
+                           new_vars = NULL) {
```
Comment (bundfussr) [inline R/derive_param_extreme_record.R:213]: I would make `new_vars` mandatory. At the moment no variables are added if `new_vars` is not specified. But this is not useful.

**Review by bundfussr — CHANGES_REQUESTED** [review]

**Comment by bundfussr** [comment]
@zdz2101 , I have resolved all my comments which you have addressed. There are two left (as Ben has ponited out) and I have added a new one.

File: R/derive_param_extreme_record.R (hunk 9 above)
[the remark this answers is not in the document]
Comment (zdz2101) [reply R/derive_param_extreme_record.R]: Updated!

File: R/derive_param_extreme_record.R (hunk 10 above)
Comment (zdz2101) [reply R/derive_param_extreme_record.R]: Updated!

**Review by bundfussr — APPROVED** [review]

**Comment by zdz2101** [comment]
@bms63 Do you want to take an additional look? Otherwise, @bundfussr might need an extra re-approval for the check templates to re-run

**Review by bundfussr — APPROVED** [review]

**Comment by bundfussr** [comment]
> Not sure why my comments were not showing up in the meeting - image

Looks like your comments are pending, i.e., you need to submit your review to make them visible for all.

File: R/derive_param_extreme_record.R [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+  assert_list_of(source_datasets, "data.frame")
+  assert_vars(by_vars, optional = TRUE)
+  assert_character_scalar(
+    mode,
+    values = c("first", "last"),
+    case_sensitive = FALSE
+  )
+  assert_varval_list(set_values_to, accept_expr = TRUE, optional = TRUE)
+
+  source_names <- names(source_datasets)
```
Comment (kaz462) [inline R/derive_param_extreme_record.R:145]: should line 145 be removed? `source_names` is not used in this file

File: R/derive_param_extreme_record.R
[the remark this answers is not in the document]
Comment (kaz462) [reply R/derive_param_extreme_record.R:1]: @bundfussr Thanks for explaining details to me! I didn't event think about defining something at a higher level, it now makes sense to me why `dataset_name` is used in `record_source`.
Regarding removing `source_datasets`, I left a comment in line 152

File: R/derive_param_extreme_record.R [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+
+  # Create Empty list to contain source datasets
+  data_list <- vector("list", length(sources))
+
+  # Evaluate the expressions contained in the sources
+  for (i in seq_along(sources)) {
+    source_dataset <- source_datasets[[sources[[i]]$dataset_name]]
```
Comment (kaz462) [inline R/derive_param_extreme_record.R:152]: I was wondering if we could do something like below to remove `source_datasets` argument, so that users don't need to link themselves.
```r
do.call("<-", list("source_dataset", as.name(sources[[1]]$dataset_name)))    
```

File: R/derive_param_extreme_record.R (hunk 11 above)
[the remark this answers is not in the document]
Comment (zdz2101) [reply R/derive_param_extreme_record.R:145]: You're right this must've been a remnant of when I was trying to get drafts/the function running off the ground

File: R/derive_param_extreme_record.R (hunk 12 above)
Comment (bundfussr) [reply R/derive_param_extreme_record.R:152]: This works only if the name of the data set coincides with the `dataset_name` field. If `dataset_name = "ae"` but the dataset to be used is `ae_ext` or `myae`, it does not work. We would force the user to rename the dataset to `ae`. This seems not user-friendly to me.

Maybe we could use it as default for the `source_datasets` argument? However, it would somehow hide the dataset. From the function call it would not be obvious that for example the `ae` dataset is used.

Comment (kaz462) [reply R/derive_param_extreme_record.R:152]: I was thinking of aligning the dataset name outside of the function call per project level settings (e.g., `myae`, `ae_ext`). 
But the current way makes sense too, I'm fine with aligning it inside the function. Thanks for the discussion!
